### PR TITLE
Feature csm encryption

### PR DESCRIPTION
**Pull request by alexemc**
# Description
Add documentation about Encryption

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/350 |

# Checklist:

- [x] Have you run a grammar and spell checks against your submission?
- [x] Have you tested the changes locally?
- [x] Have you tested whether the hyperlinks are working properly?
- [ ] Did you add the examples wherever applicable?
- [ ] Have you added high-resolution images?

### CODE REVIEW

File: content/docs/secure/encryption/deployment.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+  # When set, performance is reduced, and hard links cannot be created.
+  # See the gocryptfs documentation for more details.
+  extraArgs: []
+```
+
+| Parameter | Description | Required | Default |
+| --------- | ----------- | -------- | ------- |  
+| enabled | Enable/disable volume encryption feature. | No | false |
+| pluginName | The name of the provisioner to use when encryption is enabled. | No | "sec-isilon.dellemc.com" |
+| image | Encryption driver image name. | No | "dellemc/csm-encryption:v0.1.0" |
```
Comment (sharmilarama) [inline content/docs/secure/encryption/deployment.md:60]: this should be v1.4.0

File: content/docs/csidriver/installation/helm/isilon.md [hunk 2]
```diff
@@ -175,9 +176,12 @@ CRDs should be configured during replication prepare stage with repctl as descri
    | proxyHost | Hostname of the csm-authorization server. | No | Empty |
    | skipCertificateValidation | A boolean that enables/disables certificate validation of the csm-authorization server. | No | true |
    | **podmon**               | Podmon is an optional feature under development and tech preview. Enable this feature only after contact support for additional information.  |  -        |  -       |
-   | enabled                  | A boolean that enable/disable podmon feature. |  No      |   false   |
+   | enabled                  | A boolean that enables/disables podmon feature. |  No      |   false   |
    | image | image for podmon. | No | " " |
-
+   | **encryption** | [Encryption](../../../../secure/encryption/deployment) is an optional feature to apply encryption to CSI volumes. | - | - |
+   | enabled        | A boolean that enables/disables Encryption feature. | No | false |
+   | image | Encryption driver image name. | No | "dellemc/csm-encryption:v0.1.0" |
```
Comment (sharmilarama) [inline content/docs/csidriver/installation/helm/isilon.md:183]: this version should be v1.4.0

File: content/docs/secure/encryption/vault.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+```shell
+openssl req -newkey rsa:2048 -nodes \
+	-subj "/CN=vault-demo-server" \
+	-keyout server.key \
+	-out server.csr
+```
+
+Create server certificate signed by the CA:
+
+> Replace `10.11.12.13` below with an IP address by which Encryption can reach the Vault server.
```
Comment (sharmilarama) [inline content/docs/secure/encryption/vault.md:42]: Please mask the IP address or use a variable name.

File: content/docs/secure/encryption/_index.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+{{<table "table table-striped table-bordered table-sm">}}
+|               | PowerScale |
+|---------------|------------|
+| Storage Array | OneFS 9.0 |
+{{</table>}}
+
+## Supported CSI Drivers
+
+CSM for Encryption supports the these CSI drivers and versions:
```
Comment (rsedlock1958) [inline content/docs/secure/encryption/_index.md:85]: Delete the word "the"

File: content/docs/secure/encryption/deployment.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+the rest of the deployment process is described in the correspondent [CSI driver documentation](../../../csidriver/installation/helm).
+
+## Vault Server
+
+Hashicorp Vault must be [pre-configured](../vault) to support Encryption. The Vault server's IP address and port must be accessible 
+from the Kubernetes cluster where the CSI driver is to be deployed.
+
+## Helm Chart Values
+
+The drivers that support Encryption via Helm chart have an `encryption` block in their *values.yaml* file that looks like the following:
```
Comment (rsedlock1958) [inline content/docs/secure/encryption/deployment.md:21]: Replace "the following" with "thos"

File: content/docs/secure/encryption/deployment.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+kubectl create secret generic vault-cert -n <driver namespace> \
+	--from-file=server-ca.crt --from-file=client.crt --from-file=client.key \
+	-o yaml --dry-run=client | kubectl apply -f -
+```
+In this release, Encryption does not pick up modifications to this secret in runtime. 
+To apply new values in the secret (e.g., to update the client certificate), the CSI driver must be restarted.
+
+### ConfigMap *vault-client-conf*
+
+> Replace fields in *client.json* with your values. See the table below for available fields.
```
Comment (rsedlock1958) [inline content/docs/secure/encryption/deployment.md:133]: Replace "the table below" with "this table"

File: content/docs/secure/encryption/_index.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+| ------------- | ---------- | ------------------ |
+| CSI Driver for Dell PowerScale | [csi-powerscale](https://github.com/dell/csi-powerscale) | v2.4 + |
+{{</table>}}
+
+### PowerScale
+
+When enabling Encryption for PowerScale CSI Driver, make sure these requirements are met:
+- PowerScale CSI Driver uses root credentials for the storage array where encrypted volumes will be placed
+- OneFS NFS export configuration does not have root user mapping enabled
+- All other CSM features like Authorization, Replication, Reciliency are disabled
```
[the remark this answers is not in the document]
Comment (alexemc) [reply content/docs/secure/encryption/_index.md]: Addressed.

File: content/docs/secure/encryption/vault.md (hunk 3 above)
Comment (alexemc) [reply content/docs/secure/encryption/vault.md]: Addressed.

File: content/docs/secure/encryption/_index.md (hunk 4 above)
Comment (alexemc) [reply content/docs/secure/encryption/_index.md]: Addressed.

File: content/docs/secure/encryption/deployment.md (hunk 5 above)
Comment (alexemc) [reply content/docs/secure/encryption/deployment.md]: Addressed.

File: content/docs/secure/encryption/deployment.md (hunk 6 above)
Comment (alexemc) [reply content/docs/secure/encryption/deployment.md:133]: The sentence containing "the table below" refers to the table that does not immediately follow the sentence. Would it be appropriate to use "this" is that context?

File: content/docs/secure/encryption/deployment.md (hunk 1 above)
Comment (alexemc) [reply content/docs/secure/encryption/deployment.md:60]: After a discussion with @sharmilarama it was decided to keep v0.1.0.

File: content/docs/secure/encryption/_index.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+**Supported Vault version is 1.9.3 and newer.**
+
+Vault server (or cluster) is typically deployed in a dedicated Kubernetes cluster, but for the purpose of Encryption, it can be located anywhere.
+Even the simplest standalone single instance server with in-memory storage will suffice for testing.
+
+> **NOTE:** Properly deployed and configured Vault is crucial for security of the volumes encrypted with Encryption. 
+Please refer to the Hashicorp Vault documentation regarding recommended deployment options.
+
+> **NOTE:** Compromised Vault server or Vault storage back-end may lead to unauthorized access to the volumes encrypted with Encryption.
```
Comment (rsedlock1958) [inline content/docs/secure/encryption/_index.md:111]: This should be a Caution.

File: content/docs/secure/encryption/_index.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+> **NOTE:** Destroyed Vault storage back-end or the encryption key stored in it, will make it impossible to unlock the volume encrypted with Encryption.
```
Comment (rsedlock1958) [inline content/docs/secure/encryption/_index.md:113]: This should  be a Caution

File: content/docs/secure/encryption/deployment.md [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
```
Comment (rsedlock1958) [inline content/docs/secure/encryption/deployment.md:133]: Replace "the table below" with "this table"

File: content/docs/secure/encryption/troubleshooting.md [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+```
+NAME                              READY   STATUS    RESTARTS   AGE
+isi-controller-84f697c874-2j6d4   10/10   Running   0          16h
+isi-node-4gtwf                    4/4     Running   0          16h
+isi-node-lnzws                    4/4     Running   0          16h
+```
+
+List containers in pod `isi-node-4gtwf` with `kubectl -n <driver namespace> logs isi-node-4gtwf`. 
+Each pod has containers called `driver` which is the storage driver container and `driver-sec` which is the Encryption driver container. 
+These containers logs tend to provide the most important information, but other containers may give a hint too.
```
Comment (rsedlock1958) [inline content/docs/secure/encryption/troubleshooting.md:28]: Change "containers" to "container's"

File: content/docs/secure/encryption/vault.md [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+```shell
+openssl req -newkey rsa:2048 -nodes \
+	-subj "/CN=vault-demo-server" \
+	-keyout server.key \
+	-out server.csr
+```
+
+Create server certificate signed by the CA:
+
+> Replace `<external IP>` below with an IP address by which Encryption can reach the Vault server.
```
Comment (rsedlock1958) [inline content/docs/secure/encryption/vault.md:42]: Delete "below"

File: content/docs/secure/encryption/vault.md [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+
+vault write auth/approle/role/dea-role token_policies="dea-policy"
+```
+
+The role needs full access to the Key/Value secret engine to perform encryption key operations.
+The role only needs read access to the Key/Value secret engine config endpoint for configuration validation.
+
+#### Set role ID and secret ID to the role
+
+Role ID and secret ID are normally generated by Vault. To get the generated values, run the following commands:
```
Comment (rsedlock1958) [inline content/docs/secure/encryption/vault.md:215]: Replace "the following" with "these"

File: content/docs/secure/encryption/vault.md [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+
+> Secret ID has an expiration time after which it becomes invalid resulting in [authorization failure](../troubleshooting#expired-approle-secret-id).
+> The expiration time for new secret IDs can be set in `secret_id_ttl` parameter when [the role is created](#create-a-role) or later on using
+> `vault write auth/approle/role/dea-role/secret-id-ttl secret_id_ttl=24h`.
+
+## Token TTL Considerations
+
+Effective client token TTL is determined by the Vault server based on multiple factors which are described in the [Vault documentation](https://www.vaultproject.io/docs/concepts/tokens#token-time-to-live-periodic-tokens-and-explicit-max-ttls).
+
+With the default server settings, role level values control TTL in the following way:
```
Comment (rsedlock1958) [inline content/docs/secure/encryption/vault.md:235]: Replace "the following" with "this"

**Review by rsedlock1958 — COMMENTED** [review]
I have reviewed this. Several edits need to be made.

**Review by rsedlock1958 — DISMISSED** [review]
Looks great

File: content/docs/secure/encryption/_index.md [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+linkTitle: "Encryption"
+weight: 1
+Description: >
+  CSI Volumes Encryption
+---
+Encryption provides the capability to encrypt user data residing on volumes created by Dell CSI Drivers.
+
+> **NOTE:** This tech-preview release is not intended for use in production environment.
+
+> **NOTE:** Encryption requires a time-based license to create new encrypted volumes. Request a [trial license](../../../license) prior to deployment.
```
Comment (tdawe) [inline content/docs/secure/encryption/_index.md:12]: The license link returns a 404

Comment (alexemc) [reply content/docs/secure/encryption/_index.md]: Good finding. Fixed.

**Review by rsedlock1958 — APPROVED** [review]
I have reviewed this.  Looks good. No edits required.

**Review by sharmilarama — APPROVED** [review]

**Comment by alexemc** [comment]
Merged all commits into one and added GPG signature. No content changes.

**Review by rsedlock1958 — COMMENTED** [review]
I have reviewed this. No edits are required.

**Comment by alexemc** [comment]
Could you re-approve?

**Review by rsedlock1958 — APPROVED** [review]

**Review by sharmilarama — APPROVED** [review]

**Review by tdawe — APPROVED** [review]

**Pull request merged**